core/primitives/event_types.hpp

```diff
@@ -24,47 +27,210 @@ namespace kagome::api {
+  enum struct ChainEventType : uint32_t {
     kNewHeads = 1,
     kFinalizedHeads = 2,
     kAllHeads = 3,
     kRuntimeVersion = 4
   };
-}  // namespace kagome::primitives
 
-namespace kagome::subscription {
-  template <typename Key, typename Type, typename... Arguments>
-  class SubscriptionEngine;
+  /**
+   * - "future" - Transaction is part of the future queue.
+   * - "ready" - Transaction is part of the ready queue.
+   * - OBJECT - The transaction has been broadcast to the given peers.
+   *     "broadcast": ARRAY
+   *     STRING - PeerId.
+   * - OBJECT - Transaction has been included in block with given hash.
+   *     "inBlock": STRING - Hex-encoded hash of the block.
+   * - OBJECT - "The block this transaction was included in has been retracted.
+   *     "retracted": STRING - Hex-encoded hash of the block.
+   * - OBJECT - Maximum number of finality watchers has been reached, old
+   *    watches are being removed.
+   *    "finalityTimeout": STRING - Hex-encoded hash of the block.
+   * - OBJECT - Transaction has been finalized by GRANDPA.
+   *    "finalized": STRING - Hex-encoded hash of the block.
+   * - OBJECT - Transaction has been
+   *    replaced in the pool, by another transaction that provides the same tags
+   *    (e.g. same sender/nonce).
+   *    "usurped": STRING - Hex-encoded hash of the transaction.
+   * - "dropped" - Transaction has been dropped from the pool
+   *    because of the limit.
+   * - "invalid" - Transaction is no longer valid in the
+   *    current state.
+   */
+  enum class ExtrinsicEventType {
+    FUTURE,
+    READY,
+    BROADCAST,
+    IN_BLOCK,
+    RETRACTED,
+    FINALITY_TIMEOUT,
+    FINALIZED,
+    USURPED,
+    DROPPED,
+    INVALID
+  };
 
-  template <typename Key, typename Type, typename... Arguments>
-  class Subscriber;
-}  // namespace kagome::subscription
+  struct BroadcastEventParams {
+    std::vector<libp2p::peer::PeerId> peers;
+  };
 
-namespace kagome::subscriptions {
-  using EventsSubscriptionEngineType = subscription::SubscriptionEngine<
-      primitives::SubscriptionEventType,
-      std::shared_ptr<api::Session>,
-      boost::variant<std::reference_wrapper<boost::none_t>,
-                     std::reference_wrapper<const primitives::BlockHeader>,
-                     std::reference_wrapper<primitives::Version>>>;
-  using EventsSubscriptionEnginePtr =
-      std::shared_ptr<EventsSubscriptionEngineType>;
-  using EventsSubscribedSessionType =
-      EventsSubscriptionEngineType::SubscriberType;
-  using EventsSubscribedSessionPtr =
-      std::shared_ptr<EventsSubscribedSessionType>;
-
-  using SubscriptionEngineType =
+  struct InBlockEventParams {
+    primitives::BlockHash block;
+  };
+  struct RetractedEventParams {
+    primitives::BlockHash retracted_block;
+  };
+
+  struct FinalityTimeoutEventParams {
+    primitives::BlockHash block;
+  };
+
+  struct FinalizedEventParams {
+    primitives::BlockHash block;
+  };
```

Comment:
It seems each event call will make a copy of BlockHash. It will cause very big performance penality.